File: implementations/async-graphql/src/schema/deprecated_product.rs
use crate::schema::user::User;
use async_graphql::{Context, Object, SimpleObject};

#[derive(SimpleObject)]
pub struct DeprecatedProduct<'a> {
    sku: &'a str,
    package: &'a str,
    reason: Option<&'a str>,
    created_by: Option<User>,
}

const DEPRECATED_PRODUCT: DeprecatedProduct = DeprecatedProduct {
    sku: "apollo-federation-v1",
    package: "@apollo/federation-v1",
    reason: Some("Migrate to Federation V2"),
    created_by: None,
};

#[derive(Default)]
pub struct DeprecatedProductQuery;

#[Object]
impl DeprecatedProductQuery {
    #[graphql(deprecation = "Use product query instead")]
    pub(crate) async fn deprecated_product<'a>(
        &self,
        _ctx: &'a Context<'_>,
        sku: String,
        package: String,
    ) -> Option<&'a DeprecatedProduct<'a>> {
        if sku == DEPRECATED_PRODUCT.sku && package == DEPRECATED_PRODUCT.package {
            return Some(&DEPRECATED_PRODUCT);
        }

        None
    }

    #[graphql(entity)]
    async fn find_deprecated_product_by_sku_and_package<'a>(
        &self,
        _ctx: &'a Context<'_>,
        sku: String,
        package: String,
    ) -> Option<&'a DeprecatedProduct<'a>> {
        if sku == DEPRECATED_PRODUCT.sku && package == DEPRECATED_PRODUCT.package {
            return Some(&DEPRECATED_PRODUCT);
        }

        None
    }
}

File: implementations/graphql-java-kickstart/src/main/java/graphql/kickstart/federation/compatibility/model/DeprecatedProduct.java
package graphql.kickstart.federation.compatibility.model;

import static graphql.kickstart.federation.compatibility.model.User.DEFAULT_USER;

public record DeprecatedProduct(String sku, String pkg, String reason, User createdBy) {
    public static DeprecatedProduct DEPRECATED_PRODUCT = new DeprecatedProduct("apollo-federation-v1", "@apollo/federation-v1", "Migrate to Federation V2", DEFAULT_USER);

    public String getPackage() {
        return pkg;
    }

    public static DeprecatedProduct resolveBySkuAndPackage(String sku, String pkg) {
        if (DEPRECATED_PRODUCT.sku.equals(sku) && DEPRECATED_PRODUCT.pkg.equals(pkg))  {
            return DEPRECATED_PRODUCT;
        } else {
            return null;
        }
    }
}
